### PR TITLE
Fix expression rewrite

### PR DESCRIPTION
**Pull request by Aiee**
#### What type of PR is this?
- [x] bug
- [ ] feature
- [x] enhancement

#### What problem does this PR solve?
Issue(s) number: 
Fix https://github.com/vesoft-inc/nebula/issues/3248

Description:
This PR fixes `filterTransform()` and adds cases.

#### How do you slove it?
1. Do not transfer the filter expression if it contains 2 different tagProp expressions i.e. `v.player.name =="abc" && v.player.age == 30`
2. Fix expression overflow check. Do not rewrite the expression if it would cause an overflow.
3. Refactor `rewriteRelExpr()`
4. Return semantic error when the non-addition arithmetic expression contains strings.

#### Special notes for your reviewer, ex. impact of this fix, design document, etc:

```
(root@nebula) [nba]> match (v:player) where v.player.name+'n'=="Tim Duncann" return v
+-----------------------------------------------------+
| v                                                   |
+-----------------------------------------------------+
| ("Tim Duncan" :player{age: 42, name: "Tim Duncan"}) |
+-----------------------------------------------------+
Got 1 rows (time spent 5509/5861 us)

Fri, 31 Dec 2021 22:51:01 CST

(root@nebula) [NBA]> match (v:player) where v.player.name-'n'=="Tim Duncann" return v
`(v.player.name-"n")' is not a valid expression, can not apply `-' to `__EMPTY__' and `STRING'.

Fri, 31 Dec 2021 22:51:08 CST
```

#### Checklist:
Tests:
- [x] Unit test(positive and negative cases)
- [ ] Function test
- [ ] Perfomance test
- [ ] N/A

Affects:
- [ ] Documentation affected (Please add the label if documentation needs to be modified.)
- [ ] Incompatibility (If it breaks the compatibility, please describe it and add the label.）
- [ ] If it's needed to cherry-pick (If cherry-pick to some branches is required, please label the destination version(s).)
- [ ] Performance impacted: Consumes more CPU/Memory

#### Release notes:

Please confirm whether to be reflected in release notes and how to describe:
> ex. Fixed the bug .....

### CODE REVIEW

File: src/graph/util/ExpressionUtils.cpp [hunk 1]
```diff
@@ -413,21 +414,57 @@ Expression *ExpressionUtils::reduceUnaryNotExpr(const Expression *expr) {
...
+    if (!checkRightOperand) return false;
+
+    bool checkLeftOperand = true;
+    auto lExpr = relExpr->left();
+
+    // Left operand is arithmetical expr
+    if (lExpr->isArithmeticExpr()) {
+      auto arithmExpr = static_cast<const ArithmeticExpression *>(lExpr);
+      auto constExprList = collectAll(arithmExpr, {Expression::Kind::kConstant});
+      QueryExpressionContext ctx(nullptr);
```
Comment (nevermore3) [inline src/graph/util/ExpressionUtils.cpp:460]: redundant code

File: src/graph/util/ExpressionUtils.cpp [hunk 2]
```diff
@@ -519,21 +556,41 @@ Expression *ExpressionUtils::rewriteRelExprHelper(const Expression *expr,
...
 }
 
 StatusOr<Expression *> ExpressionUtils::filterTransform(const Expression *filter) {
-  auto rewrittenExpr = const_cast<Expression *>(filter);
+  // If the filter contains more than one different LabelAttribute expr, this filter cannot be
+  // pushed down
+  auto propExprs = ExpressionUtils::collectAll(filter, {Expression::Kind::kLabelTagProperty});
+  // Deduplicate the list
+  auto dedupPropExprsList =
+      std::unordered_set<const Expression *>(propExprs.begin(), propExprs.end());
```
Comment (nevermore3) [inline src/graph/util/ExpressionUtils.cpp:572]: should  de-duplicated the value not the pointer

File: src/graph/util/ExpressionUtils.cpp [hunk 3]
```diff
@@ -413,21 +414,57 @@ Expression *ExpressionUtils::reduceUnaryNotExpr(const Expression *expr) {
...
+    // Left operand is arithmetical expr
+    if (lExpr->isArithmeticExpr()) {
+      auto arithmExpr = static_cast<const ArithmeticExpression *>(lExpr);
+      auto constExprList = collectAll(arithmExpr, {Expression::Kind::kConstant});
+      QueryExpressionContext ctx(nullptr);
+
+      // If the arithExpr has constant expr as member that is a string, do not rewrite
+      checkLeftOperand = checkArithmExpr(arithmExpr);
+    } else if (lExpr->isRelExpr()) {  // for expressions that contain boolean literals
+      checkLeftOperand = true;
```
Comment (nevermore3) [inline src/graph/util/ExpressionUtils.cpp:465]:
```
if (lExpr->isArithmeticExpr()) {
...
}
return checkLeftOperand
```

File: src/graph/util/ExpressionUtils.cpp (hunk 2 above)
Comment (Aiee) [reply src/graph/util/ExpressionUtils.cpp:572]: Nice catch

File: src/graph/util/ExpressionUtils.cpp (hunk 1 above)
Comment (Aiee) [reply src/graph/util/ExpressionUtils.cpp:460]: How is this redundant?

Comment (nevermore3) [reply src/graph/util/ExpressionUtils.cpp:460]: where is ctx used

**Review by nevermore3 — DISMISSED** [review]

File: tests/tck/features/match/Base.feature [hunk 4]
```diff
@@ -685,3 +685,15 @@ Feature: Basic match
...
+    When executing query:
+      """
+      MATCH (v:player) WHERE v.player.name-'n'=="Tim Duncann" RETURN v
+      """
+    Then a SemanticError should be raised at runtime: `(v.player.name-"n")' is not a valid expression, can not apply `-' to `__EMPTY__' and `STRING'.
+    When executing query:
+      """
+      MATCH (v:player) WHERE v.player.name+'n'=="Tim Duncann" RETURN v
+      """
+    Then the execution should be successful
```
Comment (Shylock-Hg) [inline tests/tck/features/match/Base.feature:699]: Please check the result.

File: src/graph/util/ExpressionUtils.cpp [hunk 5]
```diff
@@ -521,21 +560,43 @@ Expression *ExpressionUtils::rewriteRelExprHelper(const Expression *expr,
...
+  std::unordered_set<std::string> dedupPropExprSet;
+  for (auto &iter : propExprs) {
+    dedupPropExprSet.emplace(iter->toString());
+  }
+  if (dedupPropExprSet.size() > 1) {
+    return const_cast<Expression *>(filter);
+  }
+
+  // Check if any overflow happen before filter tranform
+  auto initilConstFold = foldConstantExpr(filter);
```
Comment (Shylock-Hg) [inline src/graph/util/ExpressionUtils.cpp:584]: initialConstFold?

File: src/graph/util/ExpressionUtils.cpp [hunk 6]
```diff
@@ -521,21 +560,43 @@ Expression *ExpressionUtils::rewriteRelExprHelper(const Expression *expr,
...
+  NG_RETURN_IF_ERROR(initilConstFold);
+
   // Rewrite relational expression
+  auto rewrittenExpr = const_cast<Expression *>(filter);
   rewrittenExpr = rewriteRelExpr(rewrittenExpr);
```
Comment (Shylock-Hg) [inline src/graph/util/ExpressionUtils.cpp:589]: Can't rewrite folded filter directly?

Comment (Shylock-Hg) [inline src/graph/util/ExpressionUtils.cpp:589]: Can't rewrite folded filter directly?

File: tests/tck/features/match/Base.feature (hunk 4 above)
Comment (Aiee) [reply tests/tck/features/match/Base.feature:699]: The point here is to distinguish between failure and success cases, but I'll add the result later.

File: conf/nebula-graphd.conf.default [hunk 7]
```diff
@@ -84,8 +84,9 @@
 # System memory high watermark ratio, cancel the memory checking when the ratio greater than 1.0
 --system_memory_high_watermark_ratio=0.8
 
+########## metrics ##########
```
Comment (Shylock-Hg) [inline conf/nebula-graphd.conf.default:87]: Better to do it in separate PR, don't modify unrelated code.

Comment (jievince) [reply conf/nebula-graphd.conf.default:87]: emm, just a format change

File: src/graph/util/ExpressionUtils.cpp [hunk 8]
```diff
@@ -415,30 +417,67 @@ Expression *ExpressionUtils::reduceUnaryNotExpr(const Expression *expr) {
...
-  auto matcher = [](const Expression *e) -> bool {
-    if (e->isRelExpr()) {
-      auto relExpr = static_cast<const RelationalExpression *>(e);
-      if (isEvaluableExpr(relExpr->right())) {
-        return true;
+  QueryExpressionContext ctx(nullptr);
+
+  auto checkArithmExpr = [&ctx](const ArithmeticExpression *arithmExpr) -> bool {
+    auto lExpr = const_cast<Expression *>(arithmExpr->left());
+    auto rExpr = const_cast<Expression *>(arithmExpr->right());
```
Comment (jievince) [inline src/graph/util/ExpressionUtils.cpp:424]: For constant expression, you could call its method `value()` directly, so that you don't need to call const_cast.

**Review by CPWstatic — COMMENTED** [review]
Do we fix the bug that i meet in ldbc test cases?
```
        WHERE toInteger(message1.creationDate)/10000000000000   == year1
          AND toInteger(message1.creationDate)/100000000000%100 == month1
```

**Comment by Aiee** [comment]
> Do we fix the bug that i meet in ldbc test cases?
> 
> ```
>         WHERE toInteger(message1.creationDate)/10000000000000   == year1
>           AND toInteger(message1.creationDate)/100000000000%100 == month1
> ```

Yes the case you mentioned earlier has been fixed:
```
(root@nebula) [nba]> MATCH (v:player) WHERE toInteger(v.player.age)/10000000000000 == 10 AND toInteger(v.player.age)/10000000000000%100 == 1100000 RETURN v
+---+
| v |
+---+
+---+
Empty set (time spent 10065/10403 us)
```

File: src/graph/util/ExpressionUtils.cpp (hunk 8 above)
Comment (Aiee) [reply src/graph/util/ExpressionUtils.cpp:424]: Here we are still not sure about the type of `arithmExpr->left()` and `arithmExpr->right()`

**Review by CPWstatic — COMMENTED** [review]
LGTM.

File: src/graph/util/ExpressionUtils.cpp [hunk 9]
```diff
@@ -415,30 +417,67 @@ Expression *ExpressionUtils::reduceUnaryNotExpr(const Expression *expr) {
...
+
+    // If the arithExpr has constant expr as member that is a string, do not rewrite
+    if (lExpr->kind() == Expression::Kind::kConstant) {
+      if (static_cast<ConstantExpression *>(lExpr)->eval(ctx).isStr()) {
```
[the remark this answers is not in the document]
Comment (Aiee) [reply src/graph/util/ExpressionUtils.cpp]: Fixed.

**Review by CPWstatic — APPROVED** [review]

File: tests/tck/features/match/Base.IntVid.feature [hunk 10]
```diff
@@ -575,3 +575,17 @@ Feature: Basic match
       RETURN count(names)
       """
     Then a SemanticError should be raised at runtime: To get the property of the vertex in `v3.name', should use the format `var.tag.prop'
+
+  Scenario: filter is not a valid expression
+    When executing query:
+      """
+      MATCH (v:player) WHERE v.player.name-'n'=="Tim Duncann" RETURN v
+      """
+    Then a SemanticError should be raised at runtime: `(v.player.name-"n")' is not a valid expression, can not apply `-' to `__EMPTY__' and `STRING'.
```
Comment (Shylock-Hg) [inline tests/tck/features/match/Base.IntVid.feature:584]: This error message is confusing. Where is the `__EMPTY__`?

Comment (Aiee) [reply tests/tck/features/match/Base.IntVid.feature:584]: The error message will try to evaluate the expression and print its type in the message. This is to unify the error messages we currently using.

**Review by jievince — APPROVED** [review]

File: src/graph/util/ExpressionUtils.cpp [hunk 11]
```diff
@@ -415,30 +417,69 @@ Expression *ExpressionUtils::reduceUnaryNotExpr(const Expression *expr) {
...
-      if (isEvaluableExpr(relExpr->right())) {
-        return true;
+  QueryExpressionContext ctx(nullptr);
+
+  auto checkArithmExpr = [&ctx](const ArithmeticExpression *arithmExpr) -> bool {
+    auto lExpr = const_cast<Expression *>(arithmExpr->left());
+    auto rExpr = const_cast<Expression *>(arithmExpr->right());
+
+    // If the arithExpr has constant expr as member that is a string, do not rewrite
+    if (lExpr->kind() == Expression::Kind::kConstant) {
```
Comment (czpmango) [inline src/graph/util/ExpressionUtils.cpp:427]: ExpressionUtils::isEvaluable

File: src/graph/util/ExpressionUtils.cpp [hunk 12]
```diff
@@ -415,30 +417,69 @@ Expression *ExpressionUtils::reduceUnaryNotExpr(const Expression *expr) {
...
+      if (lExpr->eval(ctx).isStr()) {
+        return false;
       }
-      // TODO: To match arithmetic expression on both side
-      auto lExpr = relExpr->left();
-      if (lExpr->isArithmeticExpr()) {
-        auto arithmExpr = static_cast<const ArithmeticExpression *>(lExpr);
-        return isEvaluableExpr(arithmExpr->left()) || isEvaluableExpr(arithmExpr->right());
+    }
+    if (rExpr->kind() == Expression::Kind::kConstant) {
```
Comment (czpmango) [inline src/graph/util/ExpressionUtils.cpp:432]: Ditto.

**Review by czpmango — APPROVED** [review]

File: src/graph/util/ExpressionUtils.cpp (hunk 11 above)
Comment (Aiee) [reply src/graph/util/ExpressionUtils.cpp:427]: `"abc" or "bcd"` is a evaluable expression but is not a constant expression.
